python/ql/src/experimental/semmle/python/frameworks/Yaml.qll

```diff
@@ -0,0 +1,49 @@
+/**
+ * Provides classes modeling security-relevant aspects of the `yaml` package.
+ */
+
+private import python
+private import experimental.dataflow.DataFlow
+private import experimental.dataflow.RemoteFlowSources
+private import experimental.semmle.python.Concepts
+
+private module Yaml {
+  /** Gets a reference to the `yaml` module. */
+  DataFlow::Node yaml(DataFlow::TypeTracker t) {
+    t.start() and
+    result = DataFlow::importModule("yaml")
+    or
+    exists(DataFlow::TypeTracker t2 | result = yaml(t2).track(t2, t))
+  }
+
+  /** Gets a reference to the `yaml` module. */
+  DataFlow::Node yaml() { result = yaml(DataFlow::TypeTracker::end()) }
+
+  module yaml {
+    /** Gets a reference to the `yaml.load` function. */
+    DataFlow::Node load(DataFlow::TypeTracker t) {
+      t.start() and
+      result = DataFlow::importMember("yaml", "load")
+      or
+      t.startInAttr("load") and
+      result = yaml()
+      or
+      exists(DataFlow::TypeTracker t2 | result = yaml::load(t2).track(t2, t))
+    }
+
+    /** Gets a reference to the `yaml.load` function. */
+    DataFlow::Node load() { result = yaml::load(DataFlow::TypeTracker::end()) }
+  }
+}
+
+/**
+ * A call to `yaml.load`
+ * See https://pyyaml.org/wiki/PyYAMLDocumentation
+ */
```

Comment:
I'm not happy about marking `yaml.load` as an unsafe deserialization. After version 5.1, `yaml.load` no longer allows code execution: https://github.com/yaml/pyyaml/wiki/PyYAML-yaml.load(input)-Deprecation (so this is now a problem dependabot can handle :wink:)

There are still some attack vectors that allow code execution though, and it would be good to model them instead.

Reply:
As I read that page, the problem is still there:
> The FullLoader loader class and full_load function (the current default for load) should be avoided for now.
They do state that it will be fixed in the next release, though..